### ISSUE TITLE
An explanation on P4runtime

### ISSUE THREAD

**Issue by maxsmith303030** [p4lang/tutorials#316]
Can someone please explain the mechanism via which the controller receives the P4Runtime server details which includes the `device_id` etc as well as the mechanism via which it receives the forwarding pipeline configuration? I have gone through the tutorial files and other materials but, can't seem to find it.

**Comment by jafingerhut** [comment]
I recommend reading Section 5 "Master-Slave Arbitration and Controller Replication" of the P4Runtime Specification: https://p4.org/p4runtime/spec/master/P4Runtime-Spec.html

One of the things it says there is relevant:

"The mechanism via which the controller receives the P4Runtime server details which includes the device_id, ip and port, as well as the mechanism via which it receives the Forwarding Pipeline Config, are implementation specific and beyond the scope of this specification. Similarly, the mechanism via which the P4Runtime server receives its switch config (which notably includes the device_id) is beyond the scope of this specification. Nevertheless, if the server details or switch config are transferred via the network, it is recommended to use TLS or similar encryption and authentication mechanisms to prevent eavesdropping attacks."

In other words, the mechanisms you ask about are free to be decided upon by every network operator as they see fit.

It is possible that the Stratum project might have some specific mechanisms that it implements on doing these things, but I do not know for certain.

**Comment by maxsmith303030** [comment]
I actually did read that and my question was a  "quote" from the P4Runtime specification. I wanted to understand that and I was hoping if I could get some suggested materials on that. Thanks .

**Comment by jafingerhut** [comment]
Some suggested materials on what?

On how different network operators choose to do something that is out of scope of the public specifications?  Most network operators will keep these details strictly confidential, as proprietary trade secrets.

On how the Stratum project specifies to do these things?  I think it is best to ask in a Stratum-specific message forum, not in this Github repository.

**Comment by maxsmith303030** [comment]
Okay. I didn't know they're confidential. Thanks.

**Comment by jafingerhut** [comment]
I am not saying that I know for a fact that they are all confidential.  Stratum is open source, and _if_ they have a recommended way to do it, it should be either in its documentation, code, or both.

That said, network operators tend not to want to give out lots of information on their control infrastructure, either because they are concerned it would make them easier to attack, or that it would give away information they consider to be a strategic advantage.

**Comment by maxsmith303030** [comment]
I think I understand what you mean and, it's absolutely true when it comes to intellectual property and security. I asked the question because when I initially read it, I interpreted "beyond the scope of this specification" literally, meaning it wasn't covered in the P4 Runtime specification but, it was covered elsewhere in some material.
